pkg/git/cache.go

```diff
@@ -0,0 +1,141 @@
+package git
+
+import (
+	"crypto/sha1"
+	"encoding/hex"
+	"hash"
+	"os"
+	"path"
+	"sync"
+
+	"github.com/git-depend/git-depend/pkg/utils"
+)
+
+// Stores the path of the cache
+type Cache struct {
+	sync.Mutex
+	// root path to the cache
+	root string
+	// repositories maps the URL to the filepath
+	repositories map[string]string
+}
+
+func NewCache(path string) (*Cache, error) {
+	_, err := os.Stat(path)
+	if os.IsNotExist(err) {
+		if err := os.Mkdir(path, 0755); err != nil {
+			return nil, err
+		}
+	}
+	return &Cache{
+		root:         path,
+		repositories: make(map[string]string),
+	}, nil
+}
+
+func (cache *Cache) getNewHasher() hash.Hash {
+	return sha1.New()
+}
+
+// CloneOrUpdate figures out if we need to clone the repo or simply update.
+// Clones are done by taking the shasum of the url and using that as the directory name.
+// Returns the sha and error.
+func (cache *Cache) CloneOrUpdate(url string) (string, error) {
+	h := cache.getNewHasher()
+	h.Write([]byte(url))
+	sha := hex.EncodeToString(h.Sum(nil))
+
+	directory := path.Join(cache.root, sha)
+	_, err := os.Stat(directory)
+	// If directory exists, add it to the map.
+	// If not, create it.
+	if os.IsNotExist(err) {
+		// Create a tmp directory incase something goes wrong.
+		tmp_directory := path.Join(cache.root, "tmp", sha)
+		if err := os.RemoveAll(tmp_directory); err != nil {
+			return "", err
+		}
+		if err := os.MkdirAll(tmp_directory, 0755); err != nil {
+			return "", err
+		}
+		if err := Clone(url, tmp_directory); err != nil {
+			return "", err
+		}
+		if err := os.Rename(tmp_directory, directory); err != nil {
+			return "", err
+		}
+	} else if err != nil {
+		return "", err
+	} else if err := Update(url, directory); err != nil {
+		return "", err
+	}
+
+	cache.Lock()
+	cache.repositories[url] = sha
+	cache.Unlock()
+
+	return sha, nil
+}
+
+// CloneOrUpdateMany repositories.
+// Returns the first error encountered.
+func (cache *Cache) CloneOrUpdateMany(urls []string) ([]string, error) {
+	var wg sync.WaitGroup
+	errors := make(chan error, len(urls))
+	shas := make(chan string, len(urls))
+	set := utils.NewSet()
+
+	for _, url := range urls {
+		if exists := set.Exists(url); !exists {
+			set.Add(url)
+			wg.Add(1)
+			go func(url string) {
+				defer wg.Done()
+				s, e := cache.CloneOrUpdate(url)
+				shas <- s
+				errors <- e
+			}(url)
+		}
+	}
+	wg.Wait()
+	close(errors)
+	close(shas)
+
+	// Returns the first error encountered
+	for err := range errors {
+		if err != nil {
+			return nil, err
```

Comment:
There's a number of places where you could really benefit from returning _all_ the errors (such as missing configs for a commit). Not critical right now but good to have.

Reply:
Updated to return all errors.